extensions/community/PokiGamesSDK.json

```diff
@@ -0,0 +1,321 @@
+  "name": "PokiGamesSDK",
+  "previewIconUrl": "",
+  "shortDescription": "Extension that implements Poki SDK so that your game can be hosted on Poki.",
+  "version": "0.0.1",
+  "description": [
+    "Please refer to [Poki's guide](https://sdk.poki.com/sdk-documentation.html) to know when to call the different actions.",
+    "",
+    "To initialize the SDK, use the asynchronous action \"Load SDK\" at the beginning of your very first scene."
+  ],
+  "tags": [
+    "poki",
+    "ads",
+    "sdk"
+  ],
+  "authorIds": [
+    "tgIIofwSWVNtHsFW7AR2imu8gCj1"
+  ],
+  "dependencies": [],
+  "eventsFunctions": [
+    {
+      "fullName": "",
+      "functionType": "Action",
+      "name": "onFirstSceneLoaded",
+      "sentence": "",
+      "events": [
+        {
+          "type": "BuiltinCommonInstructions::JsCode",
+          "inlineCode": "gdjs._pokiGamesSdkExtension = {};",
+          "parameterObjects": "",
+          "useStrict": true,
+          "eventsSheetExpanded": false
+        }
+      ],
+      "parameters": [],
+      "objectGroups": []
+    },
+    {
+      "async": true,
+      "description": "Load Poki SDK.",
+      "fullName": "Load SDK",
+      "functionType": "Action",
+      "group": "Poki Games",
+      "name": "LoadSDK",
+      "sentence": "Load Poki SDK",
+      "events": [
+        {
+          "type": "BuiltinCommonInstructions::JsCode",
+          "inlineCode": [
+            "function addScript(src) {",
+            "    return new Promise((resolve, reject) => {",
+            "        const scriptElement = document.createElement('script');",
+            "",
+            "        scriptElement.setAttribute('src', src);",
+            "        scriptElement.addEventListener('load', resolve);",
+            "        scriptElement.addEventListener('error', reject);",
+            "",
+            "        document.body.appendChild(scriptElement);",
+            "    });",
+            "}",
+            "addScript('https://game-cdn.poki.com/scripts/v2/poki-sdk.js').then(() => {",
+            "    PokiSDK.init().then(() => {",
+            "        console.log(\"Poki SDK successfully initialized\");",
+            "        gdjs.evtTools.common.resolveAsyncEventsFunction(eventsFunctionContext);",
+            "    }).catch(() => {",
+            "        console.log(\"Initialized, but the user likely has adblock\");",
+            "    });",
+            "})"
```

Comment:
Generally, with JS extensions, we prefer to embed the javascript code in the event rather than load it via a script tag:

 - It does not rely on the DOM (btw, whenever you use a dom api like this, the extension must be tagged as such!)
 - It loads the script immediately allowing to use features directly after
 - It does not rely on an internet connection
 - It does not rely on another server that may be temporarily down
 - If we know a script is safe when creating the extension, we can rest assured it cannot be replaced by a nefarious version by a malicious individual

Reply:
Alright, I'll change that

Reply:
I'm not sure I will include the script here entirely: There's a first script that does :
```js
    n = "kids" === e("tag"),
    t = !!window.adBridge,
    o = "yes" === e("gdhoist"),
```
and then loads different scripts depending on those values:
```js
  var r,
    a,
    u =
      ((r = window.pokiSDKVersion || e("ab") || "v2.338.0"),
      (a = "poki-sdk-core-" + r + ".js"),
      n && (a = "poki-sdk-kids-" + r + ".js"),
      t && (a = "poki-sdk-playground-" + r + ".js"),
      o && (a = "poki-sdk-hoist-" + r + ".js"),
      "https://game-cdn.poki.com/scripts/" + r + "/" + a),
    s = document.createElement("script");
  s.setAttribute("src", u),
    s.setAttribute("type", "text/javascript"),
    s.setAttribute("crossOrigin", "anonymous"),
    (s.onload = function () {
      return i.dequeue();
    }),
    document.head.appendChild(s);
```
Do you think I should hack the first script to load the right scripts under the right conditions or should I keep the current way because:


> * It does not rely on the DOM (btw, whenever you use a dom api like this, the extension must be tagged as such!)
I'm not sure that's a problem
> * It loads the script immediately allowing to use features directly after
If I add the condition isPokiSdkLoaded, it won't be an issue
> * It does not rely on an internet connection
Poki is an online platform so it's a prerequisite
> * It does not rely on another server that may be temporarily down
I would say it's Poki's problem
> * If we know a script is safe when creating the extension, we can rest assured it cannot be replaced by a nefarious version by a malicious individual
The url seems to used Poki's CDN so I would not be too worried about this

Reply:
What would be the tag to use to declare that a DOM api is used?

Reply:
> > It does not rely on the DOM (btw, whenever you use a dom api like this, the extension must be tagged as such!)
>
> I'm not sure that's a problem

It is! A game that is published on Poki could also run for example in a server environment if it's a multiplayer game, or a porting company could also run the game on a js runtime with a native rendering library exposed to run the game on a console for example...
Code that expects the DOM APIs to be present will crash for those applications unless you create a separate version of your game without the Poki SDK... Which sucks.

> The url seems to used Poki's CDN so I would not be too worried about this

A company getting compromised is not unheward of, and we should definitely be very very careful since we are using electron with all node APIs exposed to the browser, which is a common target for malicious JS code to get a lot of bad things done. This is not just Poki's problem: due to GDevelop's negligence, this would allow all PC builds and the developer's previews to get a lot of people affected.

> If I add the condition isPokiSdkLoaded, it won't be an issue

On the contrary, you completely missed the point. Having to wait for the SDK to load with "IsSDKLoaded" is a hassle for us since we must add checks everywhere that the sdk has loaded for the game not to crash if it isn't, for the user that has to make sure not to use the SDK before it is loaded if they don't want it to no-op, create a loading screen waiting for the sdk to load, etc, and also for the player that has to wait extra time for the script to load in the background. It's a bad, error-prone experience for everyone.

> and then loads different scripts depending on those values:

Nothing stops you from embedding the different sub-scripts in the js event too :p